array/array.go

```diff
@@ -0,0 +1,131 @@
+package array
+
+import (
+	"errors"
+)
+
+type Array struct {
+	arr  []int
+	size int
+}
+
+func NewArray(capacity int) *Array {
+	return &Array{
+		arr: make([]int, capacity),
+	}
+}
+
+func (a *Array) Size() int {
+	return a.size
+}
+
+func (a *Array) Capacity() int {
+	return len(a.arr)
+}
+
+func (a *Array) IsEmpty() bool {
+	return a.Size() == 0
+}
+
+func (a *Array) ValueAt(index int) (int, error) {
+	if index < 0 || index >= a.Size() {
+		return -1, errors.New("Index is out of range")
+	}
+	return a.arr[index], nil
+}
+
+func (a *Array) Insert(item int, index int) error {
+	if index < 0 || index >= a.Capacity() {
+		return errors.New("Index is out of range")
+	}
+
+	a.checkAndShrink()
+
+	// shift array to the right
+	for i := a.size; i > index; i-- {
+		a.arr[i] = a.arr[i-1]
+	}
+
+	a.arr[index] = item
+
+	// insert at index out of size range
+	// will increase the size to index + 1
+	if index >= a.Size() {
+		a.size = index + 1
```

Comment:
This case will never happen.

Reply:
Same as my comment above, this is to jump the size of the array when inserting an element outside of size range, but inside the capacity range.